AdyenCardScanner/Sources/CardImageParser.swift

```diff
@@ -0,0 +1,236 @@
+        dateFormatter.locale = Locale.current
+        dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
+        return dateFormatter
+    }()
+
+    private var cardNumber: String?
+    private var expireDate: Date?
+
+    // MARK: - CardImageParsing
+
+    func parse(
+        image: CIImage,
+        completion: @escaping (CreditCard) -> Void
+    ) {
+        guard let transformedImage = transform(image: image) else { return }
+
+        let recognizeTextRequest = VNRecognizeTextRequest()
+        recognizeTextRequest.recognitionLevel = .accurate
+        recognizeTextRequest.usesLanguageCorrection = false
+
+        let imageRequestHandler = VNImageRequestHandler(ciImage: transformedImage, options: [:])
+        try? imageRequestHandler.perform([recognizeTextRequest])
+
+        guard let results = recognizeTextRequest.results, !results.isEmpty else {
+            return
+        }
+
+        let dispatchGroup = DispatchGroup()
+
+        dispatchGroup.enter()
+        DispatchQueue.global().async {
+            if self.cardNumber == nil {
+                self.cardNumber = self.extractCardNumber(from: results)
+            }
+            dispatchGroup.leave()
+        }
+
+        dispatchGroup.enter()
+        DispatchQueue.global().async {
+            if self.expireDate == nil {
+                self.expireDate = self.extractExpireDate(from: results)
+            }
+            dispatchGroup.leave()
+        }
+
+        dispatchGroup.notify(queue: .main) {
+            guard let cardNumber = self.cardNumber, let expireDate = self.expireDate else { return }
+            let card = CreditCard(number: cardNumber, expireDate: expireDate)
+            completion(card)
+        }
+    }
+
+    // MARK: - Private
+
+    private func transform(image: CIImage) -> CIImage? {
+        return image
+            .applyNoiseReductionFilter()?
+            .applyColorControlsFilter()?
+            .applySharpnessEnhancementFilter()
+    }
+
+    // ================== DO NOT REMOVE ==================
+    // TODO: - This is logic to grey-scale can image. We want to continue experimenting with it.
+
+    func processImageForTextRecognition(image: CIImage?, threshold: CGFloat) -> CIImage? {
+        guard let image else { return nil }
+        if let grayscaleImage = convertToGrayscale(image: image) {
+            return applyThreshold(image: grayscaleImage, threshold: threshold)
+        }
+        return nil
+    }
+
+    private func convertToGrayscale(image: CIImage) -> CIImage? {
+        let colorControlsFilter = CIFilter(name: "CIColorControls")
+        colorControlsFilter?.setValue(image, forKey: kCIInputImageKey)
+        colorControlsFilter?.setValue(0.0, forKey: kCIInputSaturationKey) // No color
+        colorControlsFilter?.setValue(1.0, forKey: kCIInputBrightnessKey) // Full brightness
+        colorControlsFilter?.setValue(1.0, forKey: kCIInputContrastKey) // Full contrast
+        return colorControlsFilter?.outputImage
+    }
+
```

Comment:
You can also move `convertToGrayscale` and `applyThreshold` to extension as fluent interface methods.

BTW: Don't forget to apply `CoreImage.CIFilterBuiltins` to them and remove optional `CIImage?`